draft-irtf-cfrg-hybrid-kems.md

```diff
@@ -537,175 +545,326 @@ and L arguments.
+  tuple (dk_PQ, dk_T) of per-constituent decapsulation keys (or pointers/handles
+  to keys).
+* The `expandDecapsulationKeyG` and `expandDecapsulationKeyK` functions are
+  replaced by the following, where `decapsToEncaps()` is a function that returns
+  the encapsulation key associated with a decapsulation key:
+
+```
+def expandDecapsulationKey(dk):
+    (dk_PQ, dkT) = dk # depending on the private key storage format
+    ek_PQ = decapsToEncaps(dk_PQ)
+    ek_T = decapsToEncaps(dk_T)
+    return (ek_PQ, ek_T, dk_PQ, dk_T)
+```
 
-Given these constituent parts, we define the QSF hybrid KEM as follows:
+These deviations have both interoperability and security impacts.
 
-~~~
-def expandDecapsulationKey(seed):
-    seed_full = PRG(seed)
-    (seed_PQ, seed_T) = split(KEM_PQ.Nseed, Group_T.Nseed, seed_full)
+From an interoperatbility point of view, the use of a second format for the
+hybrid KEM decapsulation key (other than the shared seed) introduces the risk of
+incompatibilities in cases where a private key needs to be moved from one system
+to another.
 
-    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)
-    dk_T = Group_T.RandomScalar(seed_T)
-    ek_T = Group_T.Exp(Group_T.g, dk_T)
+The security impact of this deviation is to reduce binding properties from
+MAL-BIND-P-Q to LEAK-BIND-P-Q. As discussed below, binding properties can
+address a variety of attack scenarios, including LEAK scenarios in which an
+attacker has passive access to the decapsulation key and MAL scenarios in which
+an attacker can cause the victim to use a crafted decapsulation key. The above
+hybrid KEM framework assures binding properties in the face of a LEAK attacker,
+irrespective of how key generation is done. The additional provided by the
+default "shared seed" key generation upgrades this to protection against a MAL
+attacker.
 
-    return (ek_PQ, ek_T, dk_PQ, dk_T)
+As a result, this alternative style of key generation should only be used in
+environments where implementations of component algorithms do not allow
+decapsulation keys to be imported or exported.  When decapsulation keys are
+moved around, they MUST always use the seed format.
+
+## UG Framework: Universal Combiner with a Nominal Group
 
+This framework combines a PQ KEM with a nominal group, using the universal
+combiner function.  It should be used in cases where the application wants to
+use a nominal group for the traditional component, and does not want to rely on
+the C2PRI assumption for the PQ KEM.
+
+~~~
 def DeriveKeyPair(seed):
-    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsulationKey(seed)
+    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyG(seed)
     return (concat(ek_PQ, ek_T), seed)
 
-def GenerateKeyPair():
-    seed = random(Nseed)
-    return DeriveKeyPair(seed)
+def Encaps(ek):
+    (ek_PQ, ek_T) = split(KEM_PQ.Nek, Group_T.Nelem, ek)
+    (ss_PQ, ss_T, ct_PQ, ct_T) = prepareEncapG(ek_PQ, ek_T)
+    ss_H = UniversalCombiner(ss_PQ, ss_T, ct_T, ek_T, Label))
```

Comment:
The invocation of `*Combiner` seems to be doing less than the 'prepare*'/'process' functions, I wonder if we can just inline them (as one line each)

Reply:
I don't feel strongly, but I kind of like having the indirection because (a) it highlights the commonality and (b) having a name for it highlights the security assumptions.